Explicitly check for videos playing in the viewport when computing ActivityState::IsVisuallyIdle

heycam opened the pull request:
#### 033dfee4f9751d54e30e4a0891ea98093aabe890
<pre>
Explicitly check for videos playing in the viewport when computing ActivityState::IsVisuallyIdle
<a href="https://bugs.webkit.org/show_bug.cgi?id=243393">https://bugs.webkit.org/show_bug.cgi?id=243393</a>
rdar://problem/97890582

Reviewed by Eric Carlson.

ActivityState::IsVisuallyIdle on Cocoa platforms uses notifications from the
window server to determine whether there have been no updates to the window
recently. DOM timers are throttled when the ActivityState::IsVisuallyIdle flag
is present. But due to rdar://97530095, certain videos do not cause the window
server to consider the window as being updated, and so we can end up being
IsVisuallyIdle despite the window updating, and throttling timers unexpectedly.

We use the existing viewport visibility registration that &lt;video&gt;
elements do to add a new field to MediaUsageInfo that records whether
the video is currently visible in the viewport. This state is then
consulted in WebPageProxy::updateActivityState when computing the new
value of the ActivityState::IsVisuallyIdle flag.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::isVisibleInViewportChanged):
(WebCore::MediaElementSession::updateMediaUsageIfChanged):
* Source/WebCore/platform/graphics/MediaUsageInfo.h:
(WebCore::MediaUsageInfo::operator== const):
(WebCore::MediaUsageInfo::encode const):
(WebCore::MediaUsageInfo::decode):
* Source/WebKit/UIProcess/Media/MediaUsageManager.h:
* Source/WebKit/UIProcess/Media/cocoa/MediaUsageManagerCocoa.h:
* Source/WebKit/UIProcess/Media/cocoa/MediaUsageManagerCocoa.mm:
(WebKit::MediaUsageManagerCocoa::isPlayingVideoInViewport const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateActivityState):

Canonical link: <a href="https://commits.webkit.org/253027@main">https://commits.webkit.org/253027@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2173,7 +2173,12 @@ void WebPageProxy::updateActivityState(OptionSet<ActivityState::Flag> flagsToUpd
+    bool isVisuallyIdle = pageClient().isVisuallyIdle();
+#if PLATFORM(COCOA) && !HAVE(CGS_FIX_FOR_RADAR_97530095) && ENABLE(MEDIA_USAGE)
+    if (pageClient().isViewVisible() && m_mediaUsageManager && m_mediaUsageManager->isPlayingVideoInViewport())
+        isVisuallyIdle = false;
```
There is already logic here:
```
bool WebPage::isThrottleable() const
{
    bool isActive = m_activityState.containsAny({ ActivityState::IsLoading, ActivityState::IsAudible, ActivityState::IsCapturingMedia, ActivityState::WindowIsActive });
    bool isVisuallyIdle = m_activityState.contains(ActivityState::IsVisuallyIdle);

    return m_isAppNapEnabled && !isActive && isVisuallyIdle;
}
```

To prevent throttling / app napping in certain media cases. Why can't we simply augment isThrottleable() instead of overriding the value of isVisuallyIdle?

heycam replied:
VisuallyIdle is used in a couple of other places:

* it causes us to drop the WebPageProxy::m_pageIsUserObservableCount token, in WebPageProxy::updateThrottleState
* it influences the rendering update frequency, in Page::setIsVisuallyIdleInternal

I think it's cleaner to make VisuallyIdle correct in the face of this issue with videos rather than add a new ActivityState::IsPlayingVideoInViewport flag and to check it in these places too.

cdumez replied:
Ok, I'm not objecting to the approach. Would be good if this was reviewed by a media engineer though.

eric-carlson commented on Source/WebCore/html/MediaElementSession.cpp:
```diff
@@ -298,6 +298,10 @@ void MediaElementSession::isVisibleInViewportChanged()
     if (m_element.isFullscreen() || m_element.isVisibleInViewport())
         m_elementIsHiddenUntilVisibleInViewport = false;
+
+#if PLATFORM(COCOA) && !HAVE(CGS_FIX_FOR_RADAR_97530095)
```
Minor nit: doesn't "have fix for Radar ..." imply `PLATFORM(COCOA)`?

heycam replied:
It should.

heycam replied:
But "don't have fix for Radar ..." doesn't. And we want this block to be skipped on non-Cocoa platforms.

eric-carlson replied:
I meant that Radar is Cocoa-specific so you could use `PLATFORM(COCOA)` as part of the definition of `HAVE_CGS_FIX_FOR_RADAR_97530095` now (in PlatformEnableCocoa.h ?), as you will probably do once we have a fix for rdar://97530095.

But it is a minor nit, so feel free to ignore the suggestion.

cdumez commented on Source/WebCore/html/MediaElementSession.cpp:
```diff
@@ -1273,6 +1277,9 @@ void MediaElementSession::updateMediaUsageIfChanged()
         isOutsideOfFullscreen,
         isLargeEnoughForMainContent(MediaSessionMainContentPurpose::MediaControls),
+#if PLATFORM(COCOA) && !HAVE(CGS_FIX_FOR_RADAR_97530095)
+        m_element.isVisibleInViewport()
```
Does this function get called whenever we scroll? i.e. whenever the element may have entered/exited the viewport?

heycam replied:
No it doesn't. The RenderView "is in viewport" tracking machinery, for which we already register for all video elements, only invokes its callback when it switches between in viewport / not in viewport.

eric-carlson reviewed (approved):
r=me

heycam committed:
Explicitly check for videos playing in the viewport when computing ActivityState::IsVisuallyIdle
https://bugs.webkit.org/show_bug.cgi?id=243393
rdar://problem/97890582

Reviewed by Eric Carlson.

ActivityState::IsVisuallyIdle on Cocoa platforms uses notifications from the
window server to determine whether there have been no updates to the window
recently. DOM timers are throttled when the ActivityState::IsVisuallyIdle flag
is present. But due to rdar://97530095, certain videos do not cause the window
server to consider the window as being updated, and so we can end up being
IsVisuallyIdle despite the window updating, and throttling timers unexpectedly.

We use the existing viewport visibility registration that <video>
elements do to add a new field to MediaUsageInfo that records whether
the video is currently visible in the viewport. This state is then
consulted in WebPageProxy::updateActivityState when computing the new
value of the ActivityState::IsVisuallyIdle flag.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::isVisibleInViewportChanged):
(WebCore::MediaElementSession::updateMediaUsageIfChanged):
* Source/WebCore/platform/graphics/MediaUsageInfo.h:
(WebCore::MediaUsageInfo::operator== const):
(WebCore::MediaUsageInfo::encode const):
(WebCore::MediaUsageInfo::decode):
* Source/WebKit/UIProcess/Media/MediaUsageManager.h:
* Source/WebKit/UIProcess/Media/cocoa/MediaUsageManagerCocoa.h:
* Source/WebKit/UIProcess/Media/cocoa/MediaUsageManagerCocoa.mm:
(WebKit::MediaUsageManagerCocoa::isPlayingVideoInViewport const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateActivityState):

Canonical link: https://commits.webkit.org/253027@main
